presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -79,21 +81,24 @@
     }
 
     protected final ExtendedHiveMetastore delegate;
-    private final LoadingCache<String, Optional<Database>> databaseCache;
-    private final LoadingCache<String, List<String>> databaseNamesCache;
-    private final LoadingCache<HiveTableName, Optional<Table>> tableCache;
-    private final LoadingCache<String, Optional<List<String>>> tableNamesCache;
-    private final LoadingCache<HiveTableName, PartitionStatistics> tableStatisticsCache;
-    private final LoadingCache<HivePartitionName, PartitionStatistics> partitionStatisticsCache;
-    private final LoadingCache<String, Optional<List<String>>> viewNamesCache;
-    private final LoadingCache<HivePartitionName, Optional<Partition>> partitionCache;
-    private final LoadingCache<PartitionFilter, List<String>> partitionFilterCache;
-    private final LoadingCache<HiveTableName, Optional<List<String>>> partitionNamesCache;
-    private final LoadingCache<UserTableKey, Set<HivePrivilegeInfo>> tablePrivilegesCache;
-    private final LoadingCache<String, Set<String>> rolesCache;
-    private final LoadingCache<PrestoPrincipal, Set<RoleGrant>> roleGrantsCache;
+    private final LoadingCache<KeyAndContext<String>, Optional<Database>> databaseCache;
+    private final LoadingCache<KeyAndContext<String>, List<String>> databaseNamesCache;
+    private final LoadingCache<KeyAndContext<HiveTableName>, Optional<Table>> tableCache;
+    private final LoadingCache<KeyAndContext<String>, Optional<List<String>>> tableNamesCache;
+    private final LoadingCache<KeyAndContext<HiveTableName>, PartitionStatistics> tableStatisticsCache;
+    private final LoadingCache<KeyAndContext<HivePartitionName>, PartitionStatistics> partitionStatisticsCache;
+    private final LoadingCache<KeyAndContext<String>, Optional<List<String>>> viewNamesCache;
+    private final LoadingCache<KeyAndContext<HivePartitionName>, Optional<Partition>> partitionCache;
+    private final LoadingCache<KeyAndContext<PartitionFilter>, List<String>> partitionFilterCache;
+    private final LoadingCache<KeyAndContext<HiveTableName>, Optional<List<String>>> partitionNamesCache;
+    private final LoadingCache<KeyAndContext<UserTableKey>, Set<HivePrivilegeInfo>> tablePrivilegesCache;
+    private final LoadingCache<KeyAndContext<String>, Set<String>> rolesCache;
+    private final LoadingCache<KeyAndContext<PrestoPrincipal>, Set<RoleGrant>> roleGrantsCache;
 
     private final boolean partitionVersioningEnabled;
+    private final boolean isMetastoreImpersonationEnabled;
+
+    private static final String NO_IMPERSONATION_USER = "no-impersonation-caching-user";
```

Comment:
nit: let's move the static constant to the top (as we do for other static constants in other classes)